[Glibc] Configure modulemap for target, not host

modocache opened the pull request:
<!-- Please complete this template before creating pull request. -->
#### What's in this pull request?
<!-- Description about pull request. -->

The current Glibc CMakeLists.txt uses the host machine to determine which modulemap to use. The same modulemap can't be used for all platforms because headers are available in different locations on
different platforms.

Using the host machine to determine which modulemap to configure and place at a specific path in the resource dir is fine, so long as:

1. Only one Glibc is being compiled in a single CMake invocation.
2. The target machine needs the same modulemap as the host.

https://github.com/apple/swift/pull/1442 violates both of these assumptions: the Glibc module for both Linux and Android is compiled at the same time, and the Android target can't use the Linux modulemap.

This commit instead uses the target(s) to determine which modulemap to use. The modulemap is configured and placed in an OS- and architecture-specific directory in the resource dir. The path to that modulemap is referenced by the ClangImporter (since it is no longer at a path that is automatically discovered as an implicit modulemap).

#1442 will build upon the work here by adding an Android module map:

```diff
diff --git a/stdlib/public/Glibc/CMakeLists.txt b/stdlib/public/Glibc/CMakeLists.txt
index c152ab5..f898636 100644
--- a/stdlib/public/Glibc/CMakeLists.txt
+++ b/stdlib/public/Glibc/CMakeLists.txt
@@ -18,6 +18,8 @@ foreach(SDK ${SWIFT_SDKS})
     else()
       set(GLIBC_ARCH_INCLUDE_PATH "${GLIBC_INCLUDE_PATH}")
     endif()
+  elseif("${SDK}" STREQUAL "ANDROID")
+    set(GLIBC_INCLUDE_PATH "${SWIFT_ANDROID_SDK_PATH}/usr/include")
   endif()
 
   # Verify that the location is valid.
@@ -30,6 +32,8 @@ foreach(SDK ${SWIFT_SDKS})
   set(modulemap_path "${CMAKE_CURRENT_BINARY_DIR}/${sdk}/${arch}/module.map")
   if("${SDK}" STREQUAL "FREEBSD")
     configure_file(module.freebsd.map.in "${modulemap_path}" @ONLY)
+  elseif("${SDK}" STREQUAL "ANDROID")
+    configure_file(module.android.map.in "${modulemap_path}" @ONLY)
   else()
     configure_file(module.map.in "${modulemap_path}" @ONLY)
   endif()
diff --git a/lib/ClangImporter/ClangImporter.cpp b/lib/ClangImporter/ClangImporter.cpp
index 135f575..1098d51 100644
--- a/lib/ClangImporter/ClangImporter.cpp
+++ b/lib/ClangImporter/ClangImporter.cpp
@@ -351,12 +351,15 @@ getNormalInvocationArguments(std::vector<std::string> &invocationArgStrs,
       });
     }
   } else {
    // The module map used for Glibc depends on the target we're compiling for,
    // and is not included in the resource directory with the other implicit
-    // module maps. It's at {freebsd|linux}/{arch}/glibc/module.map.
+    // module maps. It's at {android|freebsd|linux}/{arch}/glibc/module.map.
     auto GlibcModuleMapPath =
       llvm::SmallString<128>(searchPathOpts.RuntimeResourcePath);
-    if (triple.isOSFreeBSD()) {
+    if (triple.isAndroid()) {
+      llvm::sys::path::append(GlibcModuleMapPath, "android");
+    } else if (triple.isOSFreeBSD()) {
       llvm::sys::path::append(GlibcModuleMapPath, "freebsd");
     } else {
       llvm::sys::path::append(GlibcModuleMapPath, "linux");
```

#### Resolved bug number: None
<!-- If this pull request resolves any bugs from Swift bug tracker -->

* * * *

<!-- This selection should only be completed by Swift admin -->
Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please smoke test
        OS X platform               | @swift-ci Please smoke test OS X platform
        Linux platform              | @swift-ci Please smoke test Linux platform

 **Validation Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please test
        OS X platform               | @swift-ci Please test OS X platform
        Linux platform              | @swift-ci Please test Linux platform

Note: Only members of the Apple organization can trigger swift-ci.
</details>
<!-- Thank you for your contribution to Swift! -->

tkremenek commented:
@swift-ci Please smoke test

modocache commented:
Hmm, looks like a legitimate failure when running the swift-package-manager tests. I'll check it out, thanks!

Any comments on the overall approach would be very much appreciated!! :bow:

gribozavr commented:
I haven't reviewed CMake code line-by-line, but this looks like the right direction to me.

modocache commented:
Excellent, thanks @gribozavr! That's reassuring.

modocache commented:
@tkremenek @gribozavr Updated! The tests should now pass, if one of you could do me the favor of asking @swift-ci to please test.

gribozavr commented:
@swift-ci Please test

modocache commented:
Hmm, looks like some OS X errors this time. Strange, since this shouldn't affect OS X at all. I'll look into it more--thanks!

modocache commented:
Updated once again. I added a check such that `-fmodule-map-file=` is only specified if the Glibc module map actually exists. This fixes the test failures on OS X, which were caused because `swiftc -target x86_64-unknown-linux-gnu -emit-ir` was being invoked, even though the compiler was not built to support that target (and so the module map did not exist at the specified path).

This got all my tests passing on both OS X and Linux locally--fingers crossed it also works on CI!

@gribozavr, is it alright that I left a FIXME in there? Also, an alternative approach would have been to modify the tests such that `swiftc -target x86_64-unknown-linux-gnu -emit-ir` is only tested on Linux hosts. Would this be preferable?

gribozavr commented:
It would be best to keep the current extended coverage.

modocache commented:
Excellent. Well, this pull request has all the same test coverage, but can also handle the Android changes in #1442 -- whaddya think? Also, could you kick off a test?

gribozavr commented:
@swift-ci Please test

gribozavr commented:
CC @jrose-apple for Clang importer changes.

gribozavr commented on stdlib/public/Glibc/CMakeLists.txt:
```diff
@@ -1,46 +1,59 @@
-    Glibc.swift
-    Misc.c
+  Glibc.swift
+  Misc.c
```
CMake code uses 4 spaces for indentation on continuation lines.

gribozavr commented on stdlib/public/Glibc/CMakeLists.txt:
```diff
@@ -1,46 +1,59 @@
-set(output_dir "${SWIFTLIB_DIR}/glibc")
+# When cross-compiling the stdlib on Unix platforms, we'll need a separate
+# glibc for each target.
+foreach(SDK ${SWIFT_SDKS})
```
To actually make this work, you need to change this condition in the parent `CMakeLists.txt`:

```CMake
if(CMAKE_SYSTEM_NAME STREQUAL "Linux" OR CMAKE_SYSTEM_NAME STREQUAL "FreeBSD")
  add_subdirectory(Glibc)
endif()
```

modocache replied:
The parent `CMakeLists.txt` [already has those very same lines](https://github.com/apple/swift/blob/2ded9fe4447553050f08a4d1afe3c92d57b93390/stdlib/public/CMakeLists.txt#L42-L44)! And to be clear: this does actually work, both on its own when building for Linux (and I assume FreeBSD), and when #1442 is rebased on top of these changes.

modocache replied:
Oh sorry, I misread--you're saying that condition needs to change somehow?

gribozavr replied:
Correct.  Imagine trying to cross-compile from OS X to Linux.  The Glibc directory won't even be considered.

modocache replied:
Aha! And that, in fact, was the very problem with the OS X CI. Thanks for pointing this out.

gribozavr commented on stdlib/public/Glibc/CMakeLists.txt:
```diff
@@ -1,46 +1,59 @@
+# glibc for each target.
+foreach(SDK ${SWIFT_SDKS})
+  foreach(arch ${SWIFT_SDK_${SDK}_ARCHITECTURES})
+  string(TOLOWER "${SDK}" sdk)
```
This does not look right.  It should be `${SWIFT_SDK_${SDK}_LIB_SUBDIR}` IIRC.

gribozavr commented on stdlib/public/Glibc/CMakeLists.txt:
```diff
@@ -1,46 +1,59 @@
+  # Determine the location of glibc based on the target.
+  if("${SDK}" STREQUAL "LINUX" OR "${SDK}" STREQUAL "FREEBSD")
+    set(GLIBC_INCLUDE_PATH "/usr/include")
+    if(CMAKE_LIBRARY_ARCHITECTURE)
```
This is fine for now, but this mixes the host and target configuration.

gribozavr commented on stdlib/public/Glibc/CMakeLists.txt:
```diff
@@ -1,46 +1,59 @@
+  add_swift_library(swiftGlibc IS_SDK_OVERLAY
     ${sources}
-    FILE_DEPENDS copy_glibc_module "${output_dir}"
+    FILE_DEPENDS "copy_${sdk}_${arch}_glibc_module" "${output_dir}"
```
We usually use target suffixes instead.

```CMake
set(VARIANT_SUFFIX "-${SWIFT_SDK_${SDK}_LIB_SUBDIR}-${ARCH}")
```

... and append that to target names.

gribozavr commented:
CMake changes LGTM modulo comments.

jrose-apple commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -358,6 +358,39 @@ getNormalInvocationArguments(std::vector<std::string> &invocationArgStrs,
+        llvm::SmallString<128>(searchPathOpts.RuntimeResourcePath);
+    }
+    if (!GlibcModuleMapPath.empty()) {
+      if (triple.isOSFreeBSD()) {
```
Please use `swift::getPlatformNameForTriple` instead of hardcoding names here.

jrose-apple commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -358,6 +358,39 @@ getNormalInvocationArguments(std::vector<std::string> &invocationArgStrs,
+    llvm::SmallString<128> GlibcModuleMapPath;
+    if (!importerOpts.OverrideResourceDir.empty()) {
+      GlibcModuleMapPath =
+        llvm::SmallString<128>(importerOpts.OverrideResourceDir);
```
Should be no need to cast here. If there's not an implicit conversion, please use `.append(…)` instead. (Or try converting through StringRef instead of SmallString.)

jrose-apple commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -358,6 +358,39 @@ getNormalInvocationArguments(std::vector<std::string> &invocationArgStrs,
+    // The module map used for Glibc depends on the target we're compiling for,
+    // and is not included in the resource directory with the other implicit
+    // module maps. It's at {freebsd|linux}/{arch}/glibc/module.map.
+    llvm::SmallString<128> GlibcModuleMapPath;
```
Nitpick: No `llvm::` needed.

jrose-apple commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -358,6 +358,39 @@ getNormalInvocationArguments(std::vector<std::string> &invocationArgStrs,
+      GlibcModuleMapPath =
+        llvm::SmallString<128>(searchPathOpts.RuntimeResourcePath);
+    }
+    if (!GlibcModuleMapPath.empty()) {
```
I would assert this, honestly. Running without a resource directory is not a supported configuration, even when building the standard library.

modocache replied:
@jrose-apple I've added the following assert:

```cpp
SmallString<128> GlibcModuleMapPath;
if (!importerOpts.OverrideResourceDir.empty()) {
  GlibcModuleMapPath = importerOpts.OverrideResourceDir;
} else if (!searchPathOpts.RuntimeResourcePath.empty()) {
  GlibcModuleMapPath = searchPathOpts.RuntimeResourcePath;
}

// Running without a resource directory is not a supported configuration.
assert(!GlibcModuleMapPath.empty());
```

The assert fired when running `test/DebugInfo/ASTSection.swift` on Linux--it turns out it was an issue with `modulewrap_main.cpp`. I've updated that code path to properly set the runtime library directory--let me know if you agree with the changes! :bow:

jrose-apple commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -358,6 +358,39 @@ getNormalInvocationArguments(std::vector<std::string> &invocationArgStrs,
+        llvm::sys::path::append(GlibcModuleMapPath, "linux");
+      }
+      llvm::sys::path::append(
+        GlibcModuleMapPath, triple.getArchName(), "glibc", "module.map");
```
Since the module map is loaded explicitly by name, does it need to be in a subfolder, or would "glibc.modulemap" be a good enough name?

modocache replied:
That'd be awesome! I also like the idea of migrating away from the legacy "module.map" name.

jrose-apple commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -358,6 +358,39 @@ getNormalInvocationArguments(std::vector<std::string> &invocationArgStrs,
+      // Only specify the module map if that file actually exists.
+      // It may not--for example in the case that
+      // `swiftc -target x86_64-unknown-linux-gnu -emit-ir` is invoked using
+      // a Swift compiler not build for Linux targets.
```
Typo: "build" instead of "built".

jrose-apple commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -358,6 +358,39 @@ getNormalInvocationArguments(std::vector<std::string> &invocationArgStrs,
+      // a Swift compiler not build for Linux targets.
+      if (llvm::sys::fs::exists(GlibcModuleMapPath)) {
+        invocationArgStrs.insert(invocationArgStrs.end(), {
+          (Twine("-fmodule-map-file=") + GlibcModuleMapPath).str(),
```
This goes through `std::string`; it would be nice™ to avoid the allocation cost by continuing to use SmallString.

jrose-apple replied:
We also need to make sure this doesn't warn/error when the user has explicitly passed `-sdk ""` and the header files can't be found. That might just mean guarding this whole section in a check for a valid SDK, though.

modocache replied:
I suppose it's already guarded by `llvm::sys::fs::exists()`, but would you want something more tied to the validity of the SDK?

modocache commented:
@jrose-apple Thanks for the review!! I hope I've addressed all of your comments--I'm very new to C++, let alone llvm/ADT, so I can't be sure I did. :sweat_smile:

modocache commented:
@gribozavr All tests pass locally for me. Could you ask @swift-ci to test? Also, any additional feedback here?

jrose-apple commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -358,6 +358,36 @@ getNormalInvocationArguments(std::vector<std::string> &invocationArgStrs,
+    // a Swift compiler not built for Linux targets.
+    if (llvm::sys::fs::exists(GlibcModuleMapPath)) {
+      invocationArgStrs.insert(invocationArgStrs.end(), {
+        (SmallString<128>("-fmodule-map-file=") + GlibcModuleMapPath).str(),
```
I think this is *still* going to go through std::string, but maybe that's all right. I forgot this is building up a vector of std::strings anyway. Let's go back to the Twine version.

Actually, you can simplify it a bit with `push_back`. `insert` is just a trick for adding multiple arguments at once.

````c++
invocationArgStrs.push_back(
    (Twine("-fmodule-map-file=") + GlibcModuleMapPath).str());
````

modocache replied:
Thanks for the tips! Updated to use `push_back()` and `Twine`.

jrose-apple commented:
@swift-ci Please test

modocache commented:
I wonder if amending the commit to use `Twine` caused CI to abort the test run...?

jrose-apple commented:
It's probably still running, but I guess GitHub can't report status for it. I'll trigger it again.

jrose-apple commented:
@swift-ci Please test

modocache commented:
@jrose-apple @gribozavr Hooray, everything passed! Does this look good to merge?

jrose-apple commented:
Importer changes look good to me. :+1:

gribozavr commented:
CMake LGTM!

modocache commented:
Well, if everything looks fine to everyone and the CI says it's alright, I guess I've got no choice but to merge this! :grin:

Thanks again for the reviews! This will be a big help for #1442.

jrose-apple commented on stdlib/public/Glibc/CMakeLists.txt:
```diff
@@ -3,44 +3,56 @@ set(sources
-swift_install_in_component(stdlib
-    FILES "${output_dir}/module.map"
-    DESTINATION "lib/swift/glibc")
+    swift_install_in_component(stdlib
```
It looks like we've got a problem: this shows up as part of the install component even when we're not building for Linux or FreeBSD. Maybe this just needs to go in the `if` below? (Maybe *everything* needs to be guarded by that `if`?)

modocache replied:
Good call, I believe I may have misinterpreted some of the comments in the pull request review. I could place an `if("${SDK}" STREQUAL "LINUX" OR "${SDK}" STREQUAL "${FREEBSD}")` at the very beginning of the inner loop--does that sound right?

jrose-apple replied:
Maybe. Why would `"${FREEBSD}"` be a variable?

modocache replied:
I copy-pasted from five lines down, but you're absolutely right. I'll change that part, too. :sweat_smile:
